### DISCUSSION TITLE
string.EndsWith works differently? What am I missing?

### DISCUSSION THREAD

**Discussion by wstaelens** [dotnet/runtime#93961] [General] [answered]
In real life I have an asp.net core web application with Blazor server components running .net 7 (windows) **(A)**.

Calls are made to another class library (.net 7 windows) **(B)** which in turn makes calls to a class library (.net framework 4.8) **(C)**.

```(A) --> (B) --> (C)```

string is being passed down the road. Finally B makes a call to a static method in C passing a parameter *string bar*. 
<br>In the static method of library C a check happens if it is not a null-terminated string, add '\0' (or "\0")

```
   // method in class library C
   public static void Foo(string bar)
  {
     if (!bar.EndsWith("\0"))
     {
        bar += "\0";
     } else {

     }
  } 
```

when I pass in the string "abc" making the call from B (.net 7) to C (.net framework 4.8)  the ```  bar += "\0";``` **does not happen**.

![image](https://github.com/dotnet/runtime/assets/7447285/cb9def57-316e-446e-a59f-4bb896431861)

when I pass the string "abc" making the call from B (.net framwork 4.8) to C (.net framework 4.8) the ```  bar += "\0";``` **happens**.

![image](https://github.com/dotnet/runtime/assets/7447285/29653e18-aa07-427a-b5f6-b9e72100556b)

Library C makes calls to e.g. XcvData which requires a null-terminated string. Code was failing on different locations while it turned out that trivial code string.EndsWidth didn't work. 

(Checked with ```.EndsWith(char.MinValue.ToString()) )``` and   ```char c = bar.Last()```)

In Visual Studio 2022 17.7.5 Pro x64 when I hover the parameter ```string bar```  I see "abc" as value of type string. 
(.net 7.0.402 sdk - windows 10 22h2 19045.3633)

How is this string being passed/marshalled. When switching from .net 6 to .net 7 we started having exceptions, so I think something changed to strings or marshalling of strings?

What am I missing here? Why does this happen?

**--> How do I properly check if string ends with '\0' when call from B (.net 7) to C (.net framwork 4.8) is made?**

attached sample solution (.zip file) [ConsoleApp1.zip](https://github.com/dotnet/runtime/files/13162812/ConsoleApp1.zip)

**Comment by huoyaoyuan** [comment, marked as answer]
https://github.com/dotnet/runtime/issues/92025#issuecomment-1718399201

You can use `StringComparison.Ordinal` to change the behavior.

**Reply by wstaelens** [reply]
OMG
